include/rocksdb/io_status.h

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2019-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+// Copyright (c) 2019 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+//
+// An IOStatus encapsulates the result of an operation.  It may indicate
+// success, or it may indicate an error with an associated error message.
+//
+// Multiple threads can invoke const methods on an IOStatus without
+// external synchronization, but if any of the threads may call a
+// non-const method, all threads accessing the same IOStatus must use
+// external synchronization.
+
+#pragma once
+
+#include <string>
+#include "rocksdb/slice.h"
+#ifdef OS_WIN
+#include <string.h>
+#endif
+#include <cstring>
+#include "status.h"
+
+namespace rocksdb {
+
+class IOStatus : public BaseStatus<IOStatus> {
```

Comment:
I understand the need for a new IOStatus class to add all of the extensions to Status.  But would it be better to have IOStatus extend Status and possibly have a new "Code" in Status for IO errors?  Then code could treat the return from a FS operation as either a Status or IOStatus if they cared about the specific sub-codes for retrying the errors.

Reply:
If an IOStatus extends a Status, then code like:
  Status s = <fs operation> 
and 
  IOStatus s = <fs operation>
could both be valid.  As you have it defined, the Status and IOStatus are not related, and therefore not interchangeable.

Reply:
Yes, I agree this would reduce the code changes required. I'm concerned that by implementing it this way, it would be very easy for developers to ignore IOStatus. Maintaining the error handling features would be difficult. @siying Any thoughts on this?